projects/uk.ac.man.cs.llvm/src/uk/ac/man/cs/llvm/ir/module/Metadata.java

```diff
@@ -29,28 +29,476 @@
-        return ParserListener.DEFAULT;
+    protected static long unrotateSign(long u) {
+        return (u & 1) == 1 ? ~(u >> 1) : u >> 1;
     }
 
+    // https://github.com/llvm-mirror/llvm/blob/release_38/include/llvm/Bitcode/LLVMBitCodes.h#L191
     @Override
     public void record(long id, long[] args) {
         MetadataRecord record = MetadataRecord.decode(id);
         switch (record) {
+            case STRING:
+                createString(args);
+                break;
+
+            case VALUE:
+                createValue(args);
+                break;
+
+            case NODE:
+                createNode(args);
+                break;
+
+            case NAME:
+                createName(args);
+                break;
+
+            case DISTINCT_NODE:
+                createDistinctNode(args);
+                break;
+
+            case KIND:
+                createKind(args);
+                break;
+
+            case LOCATION:
+                createLocation(args);
+                break;
+
+            // OLD_NODE
+            // OLD_FN_NODE
+
+            case NAMED_NODE:
+                createNamedNode(args);
+                break;
+
+            case ATTACHMENT:
+                createAttachment(args);
+                break;
+
+            case GENERIC_DEBUG:
+                createGenericDebug(args);
+                break;
+
+            case SUBRANGE:
+                createSubrange(args);
+                break;
+
+            case ENUMERATOR:
+                createEnumerator(args);
+                break;
+
+            case BASIC_TYPE:
+                createBasicType(args);
+                break;
+
+            case FILE:
+                createFile(args);
+                break;
+
+            case SUBPROGRAM:
+                createSubprogram(args);
+                break;
+
+            case SUBROUTINE_TYPE:
+                createSubroutineType(args);
+                break;
+
+            case LEXICAL_BLOCK:
+                createLexicalBlock(args);
+                break;
+
+            case LEXICAL_BLOCK_FILE:
+                createLexicalBlockFile(args);
+                break;
+
+            case LOCAL_VAR:
+                createLocalVar(args);
+                break;
+
+            case GLOBAL_VAR:
+                createGlobalVar(args);
+                break;
+
+            case DERIVED_TYPE:
+                createDerivedType(args);
+                break;
+
+            case COMPOSITE_TYPE:
+                createCompositeType(args);
+                break;
+
+            case COMPILE_UNIT:
+                createCompileUnit(args);
+                break;
+
             default:
+                System.out.println("!" + idx + " - " + record + ": " + Arrays.toString(args));
                 break;
         }
+
+        // printMetadataDebugMsg(); // Enable to allow debugging of the metadata parser
+
+        idx++;
+    }
+
+    protected void createString(long[] args) {
+        String s = "";
```

Comment:
Can you use a `StringBuilder` here and in other places in this file? I think it makes the code easier to read (and is more efficient).